Classes/EventStore/AbstractEventSourcedRepository.php

```diff
@@ -84,28 +78,25 @@ public function findByIdentifier($identifier)
     }
 
     /**
-     * @param  AggregateRootInterface $aggregate
+     * @param AggregateRootInterface $aggregate
+     * @param int $expectedVersion
      * @return void
      */
-    public function save(AggregateRootInterface $aggregate)
+    final public function save(AggregateRootInterface $aggregate, int $expectedVersion = null)
     {
-        $streamName = $this->streamNameResolver->getStreamNameForAggregate($aggregate);
-        try {
-            $stream = $this->eventStore->get($streamName);
-        } catch (EventStreamNotFoundException $e) {
-            $stream = new EventStream();
+        if (!$aggregate instanceof EventSourcedAggregateRootInterface) {
+            throw new \InvalidArgumentException(sprintf('This repository can only save instances implementing EventSourcedAggregateRootInterface, given: "%s"', get_class($aggregate)), 1477226552);
         }
-
+        $streamName = $this->streamNameResolver->getStreamNameForAggregate($aggregate);
         $uncommittedEvents = $aggregate->pullUncommittedEvents();
-        $stream->addEvents(...$uncommittedEvents);
+        $version = $aggregate->getVersion() + 1;
+        $uncommittedEventsWithMetadata = array_map(function(EventInterface $event) use(&$version) {
+            return new EventWithMetadata($event, new EventMetadata([Metadata::VERSION => $version ++]));
+        }, $uncommittedEvents);
 
-        $this->eventStore->commit($streamName, $stream, function ($version) use ($uncommittedEvents) {
-            /** @var EventTransport $eventTransport */
-            foreach ($uncommittedEvents as $eventTransport) {
-                // @todo metadata enrichment must be done in external service, with some middleware support
-                $versionedMetadata = $eventTransport->getMetadata()->withProperty(Metadata::VERSION, $version);
-                $this->eventBus->handle($eventTransport->withMetadata($versionedMetadata));
-            }
-        });
+        if ($expectedVersion === null) {
+            $expectedVersion = $aggregate->getVersion();
```

Comment:
With this, the default (`$expectedVersion` not set) is to check concurrency based on the aggregate version.
This will prevent any events to sneak into the same stream (just like before).
With an explicit `$expectedVersion` one can implement optimistic locking like:
```php
final class UserCommandHandler implements CommandHandlerInterface
{
    public function handleRenameUser(RenameUser $command)
    {
        $user = $this->userRepository->findByIdentifier($command->getUserId());
        $user->rename($command->getNewName());
        $this->userRepository->save($user, $command->getExpectedVersion());
    }
```

By setting this to `ExpectedVersion::ANY` one can deliberately skip the concurrency check for good (this will allow for optimizations like non-transactional inserts)